File: ts/src/MaplestorySDK.ts
// Maplestory Ts SDK

import { AndroidEntity } from './entity/AndroidEntity'
import { AvatarEntity } from './entity/AvatarEntity'
import { CacheEntity } from './entity/CacheEntity'
import { CharacterEntity } from './entity/CharacterEntity'
import { ChatEntity } from './entity/ChatEntity'
import { ClusterEntity } from './entity/ClusterEntity'
import { DiffEntity } from './entity/DiffEntity'
import { Entity1Entity } from './entity/Entity1Entity'
import { GmsNewEntity } from './entity/GmsNewEntity'
import { GuildMarkEntity } from './entity/GuildMarkEntity'
import { HealthEntity } from './entity/HealthEntity'
import { ItemEntity } from './entity/ItemEntity'
import { JobEntity } from './entity/JobEntity'
import { MapEntity } from './entity/MapEntity'
import { MetricEntity } from './entity/MetricEntity'
import { MobEntity } from './entity/MobEntity'
import { MusicEntity } from './entity/MusicEntity'
import { NameEntity } from './entity/NameEntity'
import { NpcEntity } from './entity/NpcEntity'
import { NxfEntity } from './entity/NxfEntity'
import { PerformanceMetricEntity } from './entity/PerformanceMetricEntity'
import { PetEntity } from './entity/PetEntity'
import { QuestEntity } from './entity/QuestEntity'
import { SystemEntity } from './entity/SystemEntity'
import { TipEntity } from './entity/TipEntity'
import { WznEntity } from './entity/WznEntity'
import { ZMapEntity } from './entity/ZMapEntity'

export type * from './MaplestoryTypes'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { MaplestoryEntityBase } from './MaplestoryEntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class MaplestorySDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    // Add features in the resolved order (makeOptions puts an explicit
    // array order first, else defaults to test-first). Ordering matters:
    // the `test` feature installs the base mock transport and the transport
    // features (retry/cache/netsim/proxy/ratelimit) wrap whatever is current,
    // so `test` must be added before them to sit at the base of the chain.
    const featureorder = getpath(this._options, '__derived__.featureorder') || []
    for (const fname of featureorder) {
      const fopts = this._options.feature[fname] || {}
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    }

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  // Raw endpoint access is operator-controllable, like every entity op.
  // Blocking it means denying BOTH the 'direct' and 'graphql' tokens, since
  // either one reaches the same endpoint.
  async direct(fetchargs?: any) {
    if (!this._options.allow.op.includes('direct')) {
      return {
        ok: false,
        err: new Error('MaplestorySDK: direct: operation not allowed by' +
          ' SDK option allow.op value: "' + this._options.allow.op + '"'),
      }
    }

    return this._rawRequest(fetchargs)
  }


  // Ungated request path shared by direct() and graphql(), each of which
  // checks its own allow.op token first. Private, rather than a flag on
  // fetchargs: a caller-supplied marker would let anyone opt straight back
  // out of the gate by passing it.
  async _rawRequest(fetchargs?: any) {
    const utility = this._utility

    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  // Raw GraphQL access: the pressure valve that makes the generated
  // surface's deliberate omissions (per-call selection sets, typed filter
  // builders, batching, subscriptions) livable — the whole schema stays
  // reachable.
  //
  // Thin wrapper over the same prepare/fetch path `direct` uses, with the
  // one thing raw `direct` cannot do for GraphQL: a GraphQL failure rides
  // HTTP 200 as a top-level `errors` array, so status alone would report a
  // failed query as ok.
  //
  // NOTE: like `direct`, this bypasses the feature pipeline — no retry,
  // ratelimit or paging features apply.
  async graphql(query: string, variables?: any, ctrl?: any) {
    const options = this._options

    if (!options.allow.op.includes('graphql')) {
      return {
        ok: false,
        err: new Error('MaplestorySDK: graphql: operation not allowed by' +
          ' SDK option allow.op value: "' + options.allow.op + '"'),
      }
    }

    const res: any = await this._rawRequest({
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: { query, variables: variables || {} },
      ctrl,
    })

    if (res instanceof Error) {
      return res
    }

    // Errors are read BEFORE any status check: a GraphQL parse or validation
    // failure comes back as HTTP 400 carrying the standard { errors: [...] }
    // body, and the raw path represents a non-2xx as { ok: false } with no
    // err — so returning early on status would discard the server's own
    // diagnostics, which are the only useful part of that response.
    const errors = null == res.data ? undefined : res.data.errors

    if (null != errors && Array.isArray(errors) && 0 < errors.length) {
      const first = errors[0] || {}
      const err: any = new Error('MaplestorySDK: graphql: ' +
        (first.message || 'graphql error'))
      err.graphql = errors
      return { ok: false, status: res.status, headers: res.headers, err, data: res.data }
    }

    return res
  }



  // Entity access: `client.Android().list()` / `client.Android().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Android(entopts?: Record<string, any>) {
    const self = this
    return new AndroidEntity(self, entopts)
  }


  // Entity access: `client.Avatar().list()` / `client.Avatar().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Avatar(entopts?: Record<string, any>) {
    const self = this
    return new AvatarEntity(self, entopts)
  }


  // Entity access: `client.Cache().list()` / `client.Cache().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Cache(entopts?: Record<string, any>) {
    const self = this
    return new CacheEntity(self, entopts)
  }


  // Entity access: `client.Character().list()` / `client.Character().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Character(entopts?: Record<string, any>) {
    const self = this
    return new CharacterEntity(self, entopts)
  }


  // Entity access: `client.Chat().list()` / `client.Chat().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Chat(entopts?: Record<string, any>) {
    const self = this
    return new ChatEntity(self, entopts)
  }


  // Entity access: `client.Cluster().list()` / `client.Cluster().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Cluster(entopts?: Record<string, any>) {
    const self = this
    return new ClusterEntity(self, entopts)
  }


  // Entity access: `client.Diff().list()` / `client.Diff().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Diff(entopts?: Record<string, any>) {
    const self = this
    return new DiffEntity(self, entopts)
  }


  // Entity access: `client.Entity1().list()` / `client.Entity1().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Entity1(entopts?: Record<string, any>) {
    const self = this
    return new Entity1Entity(self, entopts)
  }


  // Entity access: `client.GmsNew().list()` / `client.GmsNew().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  GmsNew(entopts?: Record<string, any>) {
    const self = this
    return new GmsNewEntity(self, entopts)
  }


  // Entity access: `client.GuildMark().list()` / `client.GuildMark().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  GuildMark(entopts?: Record<string, any>) {
    const self = this
    return new GuildMarkEntity(self, entopts)
  }


  // Entity access: `client.Health().list()` / `client.Health().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Health(entopts?: Record<string, any>) {
    const self = this
    return new HealthEntity(self, entopts)
  }


  // Entity access: `client.Item().list()` / `client.Item().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Item(entopts?: Record<string, any>) {
    const self = this
    return new ItemEntity(self, entopts)
  }


  // Entity access: `client.Job().list()` / `client.Job().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Job(entopts?: Record<string, any>) {
    const self = this
    return new JobEntity(self, entopts)
  }


  // Entity access: `client.Map().list()` / `client.Map().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Map(entopts?: Record<string, any>) {
    const self = this
    return new MapEntity(self, entopts)
  }


  // Entity access: `client.Metric().list()` / `client.Metric().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Metric(entopts?: Record<string, any>) {
    const self = this
    return new MetricEntity(self, entopts)
  }


  // Entity access: `client.Mob().list()` / `client.Mob().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Mob(entopts?: Record<string, any>) {
    const self = this
    return new MobEntity(self, entopts)
  }


  // Entity access: `client.Music().list()` / `client.Music().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Music(entopts?: Record<string, any>) {
    const self = this
    return new MusicEntity(self, entopts)
  }


  // Entity access: `client.Name().list()` / `client.Name().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Name(entopts?: Record<string, any>) {
    const self = this
    return new NameEntity(self, entopts)
  }


  // Entity access: `client.Npc().list()` / `client.Npc().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Npc(entopts?: Record<string, any>) {
    const self = this
    return new NpcEntity(self, entopts)
  }


  // Entity access: `client.Nxf().list()` / `client.Nxf().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Nxf(entopts?: Record<string, any>) {
    const self = this
    return new NxfEntity(self, entopts)
  }


  // Entity access: `client.PerformanceMetric().list()` / `client.PerformanceMetric().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  PerformanceMetric(entopts?: Record<string, any>) {
    const self = this
    return new PerformanceMetricEntity(self, entopts)
  }


  // Entity access: `client.Pet().list()` / `client.Pet().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Pet(entopts?: Record<string, any>) {
    const self = this
    return new PetEntity(self, entopts)
  }


  // Entity access: `client.Quest().list()` / `client.Quest().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Quest(entopts?: Record<string, any>) {
    const self = this
    return new QuestEntity(self, entopts)
  }


  // Entity access: `client.System().list()` / `client.System().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  System(entopts?: Record<string, any>) {
    const self = this
    return new SystemEntity(self, entopts)
  }


  // Entity access: `client.Tip().list()` / `client.Tip().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Tip(entopts?: Record<string, any>) {
    const self = this
    return new TipEntity(self, entopts)
  }


  // Entity access: `client.Wzn().list()` / `client.Wzn().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Wzn(entopts?: Record<string, any>) {
    const self = this
    return new WznEntity(self, entopts)
  }


  // Entity access: `client.ZMap().list()` / `client.ZMap().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  ZMap(entopts?: Record<string, any>) {
    const self = this
    return new ZMapEntity(self, entopts)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new MaplestorySDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return MaplestorySDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'Maplestory' }
  }

  toString() {
    return 'Maplestory ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = MaplestorySDK


export {
  stdutil,
  config,

  BaseFeature,
  MaplestoryEntityBase,

  MaplestorySDK,
  SDK,
}

File: rb/Maplestory_sdk.rb
# Maplestory SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'Maplestory_types'


class MaplestorySDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = MaplestoryUtility.new
    @_utility = utility

    config = MaplestoryConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features in the resolved order (make_options puts an explicit array
    # order first, else defaults to test-first). Ordering matters: the `test`
    # feature installs the base mock transport and the transport features
    # (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
    # must be added before them to sit at the base of the chain.
    feature_opts = MaplestoryHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      featureorder = VoxgigStruct.getpath(@options, "__derived__.featureorder")
      if featureorder.is_a?(Array)
        featureorder.each do |fname|
          fopts = MaplestoryHelpers.to_map(feature_opts[fname])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, MaplestoryFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    MaplestoryUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = MaplestoryHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = MaplestoryHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = MaplestoryHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = MaplestorySpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  # Raw endpoint access is operator-controllable, like every entity op.
  # Blocking it means denying BOTH the 'direct' and 'graphql' tokens, since
  # either one reaches the same endpoint.
  def direct(fetchargs = {})
    return op_denied("direct") unless op_allowed?("direct")

    raw_request(fetchargs)
  end

  # Is this raw-access op permitted by the SDK's allow.op option?
  def op_allowed?(op)
    allow_op = VoxgigStruct.getpath(@options, "allow.op")
    allow_op.is_a?(String) && allow_op.include?(op)
  end

  def op_denied(op)
    allow_op = VoxgigStruct.getpath(@options, "allow.op")
    {
      "ok" => false,
      "err" => MaplestoryError.new(
        "#{op}_allow",
        "MaplestorySDK: #{op}: operation not allowed by" \
        " SDK option allow.op value: \"#{allow_op}\""),
    }
  end

  # Ungated request path shared by direct and graphql, each of which checks
  # its own allow.op token first. Separate, rather than a flag on fetchargs:
  # a caller-supplied marker would let anyone opt straight back out of the
  # gate by passing it.
  def raw_request(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue MaplestoryError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = MaplestoryHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = MaplestoryHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end

  # Raw GraphQL access: the pressure valve that makes the generated surface's
  # deliberate omissions (per-call selection sets, typed filter builders,
  # batching, subscriptions) livable — the whole schema stays reachable.
  #
  # Thin wrapper over the same prepare/fetch path direct uses, with the one
  # thing raw direct cannot do for GraphQL: a GraphQL failure rides HTTP 200
  # as a top-level `errors` array, so status alone would report a failed
  # query as ok.
  #
  # NOTE: like direct, this bypasses the feature pipeline — no retry,
  # ratelimit or paging features apply.
  def graphql(query, variables = nil, ctrl = nil)
    return op_denied("graphql") unless op_allowed?("graphql")

    res = raw_request({
      "method" => "POST",
      "headers" => { "content-type" => "application/json" },
      "body" => { "query" => query, "variables" => variables || {} },
      "ctrl" => ctrl || {},
    })

    # Errors are read BEFORE any status check: a GraphQL parse or validation
    # failure comes back as HTTP 400 carrying the standard { errors: [...] }
    # body, and the raw path represents a non-2xx as ok:false with no err —
    # so returning early on status would discard the server's own
    # diagnostics, which are the only useful part of that response.
    errors = VoxgigStruct.getpath(res, "data.errors")

    if errors.is_a?(Array) && !errors.empty?
      first = errors[0].is_a?(Hash) ? errors[0] : {}
      msg = first["message"]
      msg = "graphql error" if msg.nil? || msg.to_s.empty?
      res["ok"] = false
      res["err"] = MaplestoryError.new(
        "graphql_error", "MaplestorySDK: graphql: #{msg}")
      res["graphql"] = errors
    end

    res
  end


  # Canonical facade: client.Android.list / client.Android.load({ "id" => ... })
  def Android(data = nil)
    require_relative 'entity/android_entity'
    AndroidEntity.new(self, data)
  end


  # Canonical facade: client.Avatar.list / client.Avatar.load({ "id" => ... })
  def Avatar(data = nil)
    require_relative 'entity/avatar_entity'
    AvatarEntity.new(self, data)
  end


  # Canonical facade: client.Cache.list / client.Cache.load({ "id" => ... })
  def Cache(data = nil)
    require_relative 'entity/cache_entity'
    CacheEntity.new(self, data)
  end


  # Canonical facade: client.Character.list / client.Character.load({ "id" => ... })
  def Character(data = nil)
    require_relative 'entity/character_entity'
    CharacterEntity.new(self, data)
  end


  # Canonical facade: client.Chat.list / client.Chat.load({ "id" => ... })
  def Chat(data = nil)
    require_relative 'entity/chat_entity'
    ChatEntity.new(self, data)
  end


  # Canonical facade: client.Cluster.list / client.Cluster.load({ "id" => ... })
  def Cluster(data = nil)
    require_relative 'entity/cluster_entity'
    ClusterEntity.new(self, data)
  end


  # Canonical facade: client.Diff.list / client.Diff.load({ "id" => ... })
  def Diff(data = nil)
    require_relative 'entity/diff_entity'
    DiffEntity.new(self, data)
  end


  # Canonical facade: client.Entity1.list / client.Entity1.load({ "id" => ... })
  def Entity1(data = nil)
    require_relative 'entity/entity1_entity'
    Entity1Entity.new(self, data)
  end


  # Canonical facade: client.GmsNew.list / client.GmsNew.load({ "id" => ... })
  def GmsNew(data = nil)
    require_relative 'entity/gms_new_entity'
    GmsNewEntity.new(self, data)
  end


  # Canonical facade: client.GuildMark.list / client.GuildMark.load({ "id" => ... })
  def GuildMark(data = nil)
    require_relative 'entity/guild_mark_entity'
    GuildMarkEntity.new(self, data)
  end


  # Canonical facade: client.Health.list / client.Health.load({ "id" => ... })
  def Health(data = nil)
    require_relative 'entity/health_entity'
    HealthEntity.new(self, data)
  end


  # Canonical facade: client.Item.list / client.Item.load({ "id" => ... })
  def Item(data = nil)
    require_relative 'entity/item_entity'
    ItemEntity.new(self, data)
  end


  # Canonical facade: client.Job.list / client.Job.load({ "id" => ... })
  def Job(data = nil)
    require_relative 'entity/job_entity'
    JobEntity.new(self, data)
  end


  # Canonical facade: client.Map.list / client.Map.load({ "id" => ... })
  def Map(data = nil)
    require_relative 'entity/map_entity'
    MapEntity.new(self, data)
  end


  # Canonical facade: client.Metric.list / client.Metric.load({ "id" => ... })
  def Metric(data = nil)
    require_relative 'entity/metric_entity'
    MetricEntity.new(self, data)
  end


  # Canonical facade: client.Mob.list / client.Mob.load({ "id" => ... })
  def Mob(data = nil)
    require_relative 'entity/mob_entity'
    MobEntity.new(self, data)
  end


  # Canonical facade: client.Music.list / client.Music.load({ "id" => ... })
  def Music(data = nil)
    require_relative 'entity/music_entity'
    MusicEntity.new(self, data)
  end


  # Canonical facade: client.Name.list / client.Name.load({ "id" => ... })
  def Name(data = nil)
    require_relative 'entity/name_entity'
    NameEntity.new(self, data)
  end


  # Canonical facade: client.Npc.list / client.Npc.load({ "id" => ... })
  def Npc(data = nil)
    require_relative 'entity/npc_entity'
    NpcEntity.new(self, data)
  end


  # Canonical facade: client.Nxf.list / client.Nxf.load({ "id" => ... })
  def Nxf(data = nil)
    require_relative 'entity/nxf_entity'
    NxfEntity.new(self, data)
  end


  # Canonical facade: client.PerformanceMetric.list / client.PerformanceMetric.load({ "id" => ... })
  def PerformanceMetric(data = nil)
    require_relative 'entity/performance_metric_entity'
    PerformanceMetricEntity.new(self, data)
  end


  # Canonical facade: client.Pet.list / client.Pet.load({ "id" => ... })
  def Pet(data = nil)
    require_relative 'entity/pet_entity'
    PetEntity.new(self, data)
  end


  # Canonical facade: client.Quest.list / client.Quest.load({ "id" => ... })
  def Quest(data = nil)
    require_relative 'entity/quest_entity'
    QuestEntity.new(self, data)
  end


  # Canonical facade: client.System.list / client.System.load({ "id" => ... })
  def System(data = nil)
    require_relative 'entity/system_entity'
    SystemEntity.new(self, data)
  end


  # Canonical facade: client.Tip.list / client.Tip.load({ "id" => ... })
  def Tip(data = nil)
    require_relative 'entity/tip_entity'
    TipEntity.new(self, data)
  end


  # Canonical facade: client.Wzn.list / client.Wzn.load({ "id" => ... })
  def Wzn(data = nil)
    require_relative 'entity/wzn_entity'
    WznEntity.new(self, data)
  end


  # Canonical facade: client.ZMap.list / client.ZMap.load({ "id" => ... })
  def ZMap(data = nil)
    require_relative 'entity/z_map_entity'
    ZMapEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = MaplestorySDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end
